directorywatcher/directorywatcher.go

```diff
@@ -0,0 +1,288 @@
+				func() {
+					if isDir {
+						watcher.Add(ev.Name)
+					} else {
+						f, err := limitopen.OpenWithLimit(ev.Name, softLimit, hardLimit)
+						if err != nil {
+							logger.Log(context.Background(), "directorywatcher: I/O error: "+err.Error())
+							return
+						}
+						defer f.Close()
+						d, err := parser(f)
+						if err != nil {
+							logger.Log(context.Background(), "directorywatcher: parser error: "+err.Error())
+						} else {
+							data := r.data.Load()
+							data, err = add(data, d)
+							if err != nil {
+								logger.Log(context.Background(), "directorywatcher: add file error: "+err.Error())
+								return
+							}
+							r.data.Store(data)
+						}
+					}
+				}()
+			case fsnotify.Rename, fsnotify.Remove: // remove
+				// Wrap with an anonymous function to make sure that defer works.
+				func() {
+					// remove data related to file
+					data := r.data.Load()
+					data, err := remove(data, ev.Name)
+					if err != nil {
+						logger.Log(context.Background(), "directorywatcher: remove file error: "+err.Error())
+						return
+					}
+					r.data.Store(data)
+				}()
+			case fsnotify.Write, fsnotify.Chmod: //parse
+				// Wrap with an anonymous function to make sure that defer works.
+				func() {
+					if isDir {
+						// do nothing
+					} else {
+						f, err := limitopen.OpenWithLimit(ev.Name, softLimit, hardLimit)
+						if err != nil {
+							logger.Log(context.Background(), "directorywatcher: I/O error: "+err.Error())
+							return
+						}
+						defer f.Close()
+						d, err := parser(f)
+						if err != nil {
+							logger.Log(context.Background(), "directorywatcher: parser error: "+err.Error())
+						} else {
+							data := r.data.Load()
+							data, err = add(data, d)
+							if err != nil {
+								logger.Log(context.Background(), "directorywatcher: add file error: "+err.Error())
+								return
+							}
+							r.data.Store(data)
+						}
+					}
+				}()
+			}
+		}
+	}
+}
+
+var (
+	_ DirectoryWatcher = (*Result)(nil)
+)
+
+// New initializes a directorywatcher designed for recursivly
+// looking through a directory instead of a file
+func New(ctx context.Context, cfg Config) (*Result, error) {
+	limit := cfg.MaxFileSize
+	if limit <= 0 {
+		limit = filewatcher.DefaultMaxFileSize
+	}
+	hardLimit := limit * filewatcher.HardLimitMultiplier
+
+	var watcher *fsnotify.Watcher
+	watcher, err := fsnotify.NewWatcher()
+	if err != nil {
+		return nil, err
+	}
+
+	var d interface{}
+	res := &Result{}
+	// res.data.Store(Folder{
+	// 	Files: make(map[string]interface{}),
+	// })
+	// Need to walk recursively because the watcher
+	// doesnt support recursion by itself
+	dirPath := filepath.Clean(cfg.Path)
+
+	isDir, err := isDirectory(dirPath)
+	if !isDir {
+		return nil, fmt.Errorf("directorywatcher: %q is not a directory", dirPath)
+	} else if err != nil {
+		return nil, err
+	}
+
+	err = filepath.WalkDir(dirPath, func(path string, info fs.DirEntry, err error) error {
```

Comment:
There are enough nested functions here that you probably want to factor at least some of this out into helper methods.  The amount of possible local state and closure capturing is really high, which makes it hard to read and reason about.